File: apps/api/app/rutas/clickup.py
"""Rutas del conector ClickUp (listas + OAuth por empresa, Spec 009).

`GET /clickup/listas` alimenta el SELECTOR de lista destino en la pantalla de Tareas:
devuelve las listas reales del usuario (planas) para que ELIJA dónde crear las tareas
de la propuesta (decisión de producto: la lista NO va hardcodeada, es por-empresa como
un conector). Requiere auth (la empresa sale del JWT).

* Sin token configurado → 200 con `[]` (el front muestra "conecta ClickUp"); NO se
  llama a ClickUp.
* Si ClickUp falla (token inválido, red, 5xx) → 502 (servicio externo caído).

`POST /clickup/verificar` ejecuta la lógica de auto-heal de forma EXPLÍCITA: el front
lo llama solo cuando el usuario quiere verificar/reparar la conexión ClickUp. El GET
queda puro (sin side effects, RFC 7231 §4.2.1).

El OAuth por empresa (iniciar/callback/estado/desconectar) ESPEJA el de Gmail
(`rutas/integraciones.py`): el token se resuelve por-tenant y vive SÓLO en el backend
(Secret Manager); jamás viaja al front (CA5). El endpoint que CREA las tareas vive en
`rutas/solicitudes.py` (cuelga de una solicitud).
"""
import secrets
from uuid import UUID, uuid4

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import RedirectResponse

from app.dependencias import (
    obtener_almacen_estado_oauth,
    obtener_almacen_secretos,
    obtener_cliente_clickup,
    obtener_cliente_oauth_clickup,
    obtener_config_oauth_clickup,
    obtener_empresa_actual,
    obtener_repositorio_integraciones,
    obtener_repositorio_integraciones_servicio,
)
from app.esquemas import (
    EstadoClickUp,
    ListaClickUpSalida,
    ResultadoVerificacionClickUp,
    UrlConsentimiento,
)
from app.repositorios.integraciones import Integracion
from app.servicios.oauth_clickup import ConfigOAuthClickUp, construir_url_consentimiento

router = APIRouter(prefix="/clickup", tags=["clickup"])


# ── Helper de auto-heal (extraído del GET, ahora reutilizable) ────────────────

async def verificar_integracion_clickup(
    clickup,
    repo,
    empresa_id: UUID,
) -> ResultadoVerificacionClickUp:
    """Verifica la integración ClickUp de la empresa y repara el estado si es necesario.

    - Sin token → no hay integración, nada que verificar.
    - Llama a ClickUp (/listas) para probar que el token funciona.
    - Si funciona y la integración venía 'reconectar', la repara a 'conectado'.
    - Si ClickUp falla (token inválido), marca 'reconectar'.

    Retorna el estado resultante + un mensaje descriptivo."""
    if not clickup.tiene_token:
        return ResultadoVerificacionClickUp(
            estado="sin_conectar",
            mensaje="La empresa no tiene integración ClickUp configurada.",
        )

    try:
        await clickup.listar_listas()
    except Exception:
        # El token no sirve: marcar 'reconectar' si no estaba ya.
        integracion = await repo.obtener_por_empresa_y_proveedor(empresa_id, "clickup")
        if integracion is not None and integracion.estado != "reconectar":
            await repo.marcar_estado(empresa_id, "reconectar", "clickup")
        return ResultadoVerificacionClickUp(
            estado="reconectar",
            mensaje="El token de ClickUp no es válido. Se requiere reconexión.",
        )

    # Token funciona: restaurar a 'conectado' si venía en otro estado.
    integracion = await repo.obtener_por_empresa_y_proveedor(empresa_id, "clickup")
    if integracion is not None and integracion.estado != "conectado":
        await repo.marcar_estado(empresa_id, "conectado", "clickup")
        return ResultadoVerificacionClickUp(
            estado="conectado",
            mensaje="La integración fue reparada automáticamente (auto-heal).",
        )
    return ResultadoVerificacionClickUp(
        estado="conectado",
        mensaje="La integración ClickUp funciona correctamente.",
    )


@router.get("/listas", response_model=list[ListaClickUpSalida])
async def listar_listas(
    clickup=Depends(obtener_cliente_clickup),
) -> list[ListaClickUpSalida]:
    """Listas reales del ClickUp del usuario (planas) para el selector de destino.

    Endpoint PURO (sin side effects): solo lee y devuelve listas. No muta el estado
    de la integración. Para verificar/reparar la conexión, usar POST /clickup/verificar.
    """
    # Sin token: la empresa aún no conectó ClickUp → [] (el front muestra el aviso).
    if not clickup.tiene_token:
        return []
    try:
        listas = await clickup.listar_listas()
    except Exception as exc:
        # ClickUp es un servicio externo: si cae, es un 502 (no un 500 crudo).
        raise HTTPException(
            status_code=502, detail="El servicio de ClickUp no está disponible"
        ) from exc

    return [
        ListaClickUpSalida(id=l.id, nombre=l.nombre, espacio=l.espacio)
        for l in listas
    ]


@router.post("/verificar", response_model=ResultadoVerificacionClickUp)
async def verificar_clickup(
    clickup=Depends(obtener_cliente_clickup),
    repo=Depends(obtener_repositorio_integraciones),
    empresa_id: UUID = Depends(obtener_empresa_actual),
) -> ResultadoVerificacionClickUp:
    """Verifica y repara la integración ClickUp de la empresa (auto-heal explícito).

    El front llama a este endpoint cuando el usuario quiere verificar la conexión
    (botón "Verificar conexión"). Es un POST porque PUEDE mutar el estado de la
    integración (cambiar 'reconectar' → 'conectado' o viceversa).

    NOTA para el frontend: este endpoint REEMPLAZA el auto-heal implícito que antes
    vivía en GET /clickup/listas. El GET ahora es puro (sin side effects). Si el front
    detecta un 502 al cargar listas, puede ofrecer al usuario llamar a este endpoint
    para diagnosticar la conexión."""
    return await verificar_integracion_clickup(clickup, repo, empresa_id)


@router.get("/estado", response_model=EstadoClickUp)
async def estado_clickup(
    repo=Depends(obtener_repositorio_integraciones),
    empresa_id: UUID = Depends(obtener_empresa_actual),
) -> EstadoClickUp:
    """Estado del conector ClickUp de la empresa, por EXISTENCIA de la integración
    (aclaración #5: barato, sin llamar a ClickUp). Sin integración → todo null
    ("Sin conectar", CA1). El conteo de listas es una llamada viva aparte
    (`/clickup/listas`). Nunca incluye tokens (CA5, garantizado por el `response_model`)."""
    integracion = await repo.obtener_por_empresa_y_proveedor(empresa_id, "clickup")
    if integracion is None:
        return EstadoClickUp()
    return EstadoClickUp(
        proveedor=integracion.proveedor, estado=integracion.estado
    )


@router.post("/iniciar", response_model=UrlConsentimiento)
async def iniciar_clickup(
    config: ConfigOAuthClickUp = Depends(obtener_config_oauth_clickup),
    almacen=Depends(obtener_almacen_estado_oauth),
    empresa_id: UUID = Depends(obtener_empresa_actual),
) -> UrlConsentimiento:
    """Arranca el consentimiento de ClickUp: genera un `state` anti-CSRF ligado a la
    empresa, lo registra y devuelve la URL de ClickUp a la que el front redirige.

    No toca ClickUp: sólo construye la URL (client_id + redirect_uri + state). El canje
    ocurre en el callback (CA2)."""
    state = secrets.token_urlsafe(32)
    await almacen.guardar(state, empresa_id)
    return UrlConsentimiento(url=construir_url_consentimiento(config, state))


@router.get("/callback")
async def callback_clickup(
    code: str,
    state: str,
    config: ConfigOAuthClickUp = Depends(obtener_config_oauth_clickup),
    almacen_estado=Depends(obtener_almacen_estado_oauth),
    oauth=Depends(obtener_cliente_oauth_clickup),
    secretos=Depends(obtener_almacen_secretos),
    repo=Depends(obtener_repositorio_integraciones_servicio),
) -> RedirectResponse:
    """Cierra el consentimiento: valida el `state` (anti-CSRF), canjea el `code` por el
    access token, lo guarda como secreto y persiste la integración clickup conectada;
    luego redirige al front. El token nunca sale en la respuesta (CA5).

    La empresa se toma del `state` validado (no del JWT): es el navegador el que vuelve
    de ClickUp (un redirect, sin header `Authorization`), y el `state` es lo que liga
    ese retorno a su empresa. Por eso el repo es el de SERVICIO (service role): no hay
    JWT que active la RLS, así que escribimos con `empresa_id` EXPLÍCITO del state
    (jamás se infiere → no cruza tenants)."""
    empresa_id = await almacen_estado.consumir(state)
    if empresa_id is None:
        raise HTTPException(status_code=400, detail="state inválido o expirado")

    credenciales = await oauth.canjear_codigo(code)
    token_ref = await secretos.guardar(
        f"clickup-token-{empresa_id}", credenciales.access_token
    )
    await repo.guardar(
        Integracion(
            id=uuid4(),
            empresa_id=empresa_id,  # EXPLÍCITO, del state (no inferido)
            proveedor="clickup",
            token_ref=token_ref,
            casilla=None,
            cursor=None,
            estado="conectado",
        )
    )
    return RedirectResponse(url=config.url_post_conexion, status_code=302)


@router.delete("", status_code=204)
async def desconectar_clickup(
    repo=Depends(obtener_repositorio_integraciones),
    secretos=Depends(obtener_almacen_secretos),
    empresa_id: UUID = Depends(obtener_empresa_actual),
) -> None:
    """Desconecta ClickUp de la empresa (el botón "Cambiar"): borra el secreto (access
    token) y elimina SÓLO la fila clickup de `integraciones` (no toca gmail). Idempotente:
    si no hay integración clickup, responde 204 igual (CA7)."""
    integracion = await repo.obtener_por_empresa_y_proveedor(empresa_id, "clickup")
    if integracion is not None:
        await secretos.borrar(integracion.token_ref)
        await repo.eliminar_por_proveedor(empresa_id, "clickup")

File: apps/web/src/api/clickup.ts
// Cliente del conector ClickUp (estado del conector + envío de tareas).
//
// IMPORTANTE (regla de oro #3): el token de ClickUp vive SOLO en el backend; el front
// nunca lo toca. Aquí el front habla con nuestro FastAPI, que a su vez habla con la API
// de ClickUp. Operaciones:
//   * `obtenerEstadoClickup()` → ¿la EMPRESA tiene ClickUp conectado? (panel de
//     Configuración, espejo de Gmail). Sin integración → "Sin conectar"; con token
//     revocado/401 el backend marca "reconectar" (CA6). Ante fallo cae a desconectado.
//   * `iniciarConexionClickup()` → URL de consentimiento OAuth (el front la navega; el
//     `state` anti-CSRF lo fija el backend, el front NUNCA inventa la URL).
//   * `desconectarClickup()` → el "Cambiar": borra integración + secreto en el backend.
//   * `listarListasClickUp()` → puebla el SELECTOR de lista destino (las listas reales
//     del usuario). Sin token configurado el backend devuelve [] → el front muestra el
//     aviso "Conecta ClickUp". Ante cualquier fallo, también [] (no rompe la pantalla).
//   * `enviarTareasAClickUp(solicitudId, listaId)` → crea las tareas de la propuesta en
//     la lista elegida y devuelve `{creadas}`; ante error, `null` (igual que solicitudes.ts).
//
// El token de ClickUp JAMÁS viaja al frontend (CA5): el backend lo omite del
// response_model y estos clientes solo mapean proveedor/estado.

import { cabecerasAuthAsync } from './auth'

const API_BASE = (import.meta.env.VITE_API_URL as string | undefined) ?? '/api'

/** Estado del conector ClickUp de la empresa (GET /clickup/estado). Sin token. */
export interface EstadoClickup {
  proveedor: 'clickup' | null
  estado: 'conectado' | 'reconectar' | null
}

/** ClickUp sin conectar (también es el fallback ante error del backend). */
export const ESTADO_CLICKUP_DESCONECTADO: EstadoClickup = {
  proveedor: null,
  estado: null,
}

/** Una lista de ClickUp para el selector de destino (forma plana del backend). */
export interface ListaClickUp {
  id: string
  nombre: string
  espacio: string
}

/**
 * Lee el estado del conector ClickUp de la empresa autenticada (por-tenant).
 * Espejo de `obtenerEstadoCorreo`: con integración → {proveedor:'clickup', estado};
 * sin integración → todo null ("Sin conectar"); token revocado/401 → estado
 * 'reconectar' (CA6). Si el backend no responde, cae a "desconectado" (no rompe el
 * panel). Solo mapea proveedor/estado: el token nunca se expone (CA5).
 */
export async function obtenerEstadoClickup(): Promise<EstadoClickup> {
  try {
    const r = await fetch(`${API_BASE}/clickup/estado`, { headers: await cabecerasAuthAsync() })
    if (!r.ok) throw new Error(`backend respondió ${r.status}`)
    const data = (await r.json()) as Partial<EstadoClickup>
    return {
      proveedor: data.proveedor ?? null,
      estado: data.estado ?? null,
    }
  } catch {
    return ESTADO_CLICKUP_DESCONECTADO
  }
}

/**
 * Pide al backend la URL de consentimiento de ClickUp para conectar el conector.
 * Devuelve la `url` EXACTA del backend: el cliente nunca inventa la URL de OAuth
 * (el `state` anti-CSRF lo fija el servidor). Si el backend falla, propaga el error
 * para que la UI no navegue a una URL inventada (igual que Gmail).
 */
export async function iniciarConexionClickup(): Promise<string> {
  const r = await fetch(`${API_BASE}/clickup/iniciar`, {
    method: 'POST',
    headers: await cabecerasAuthAsync(),
  })
  if (!r.ok) throw new Error(`backend respondió ${r.status}`)
  const data = (await r.json()) as { url: string }
  return data.url
}

/**
 * Desconecta ClickUp de la empresa (el "Cambiar"). Borra integración + secreto en el
 * backend. Ante fallo de red no propaga (igual que `desconectarCorreo`): el panel
 * vuelve a "Sin conectar" en la UI sin romperse.
 */
export async function desconectarClickup(): Promise<void> {
  try {
    await fetch(`${API_BASE}/clickup`, {
      method: 'DELETE',
      headers: await cabecerasAuthAsync(),
    })
  } catch {
    // Red caída: no rompemos el flujo de "Cambiar".
  }
}

/**
 * Lee las listas reales del ClickUp de la empresa desde el backend. Si no hay token
 * configurado el backend responde [] (el front muestra "conecta ClickUp"); ante
 * cualquier otro fallo (502, red caída) también devuelve [] para no romper la UI.
 */
export async function listarListasClickUp(): Promise<ListaClickUp[]> {
  try {
    const r = await fetch(`${API_BASE}/clickup/listas`, { headers: await cabecerasAuthAsync() })
    if (!r.ok) throw new Error(`backend respondió ${r.status}`)
    const data = await r.json()
    // Defensa: si el backend devolviera algo inesperado, no rompemos el selector.
    return Array.isArray(data) ? (data as ListaClickUp[]) : []
  } catch {
    return []
  }
}

/**
 * Crea las tareas de la propuesta como tareas reales en ClickUp, en la lista elegida.
 * `listaId` vacío → no se manda y el backend usa su lista por defecto (fallback).
 * `asignados` (opcional, 0006) es el mapa nombre_de_tarea → persona asignada que el
 * usuario eligió en el selector de la pantalla de Tareas; viaja en el cuerpo JSON y el
 * backend lo pone en la descripción de cada tarea. Devuelve `{creadas: N}` si todo
 * salió bien, o `null` ante error (sin lista → 400, sin propuesta → 404, ClickUp caído
 * → 502, red caída) — sin romper la UI.
 */
export async function enviarTareasAClickUp(
  solicitudId: string,
  listaId: string,
  asignados: Record<string, string> = {},
): Promise<{ creadas: number } | null> {
  try {
    const qs = listaId ? `?lista_id=${encodeURIComponent(listaId)}` : ''
    const r = await fetch(`${API_BASE}/solicitudes/${solicitudId}/tareas/clickup${qs}`, {
      method: 'POST',
      // Content-Type explícito: mandamos JSON con el mapa de asignaciones (el backend
      // lo trata como opcional, así que {} es un cuerpo válido = comportamiento actual).
      headers: { ...(await cabecerasAuthAsync()), 'Content-Type': 'application/json' },
      body: JSON.stringify({ asignados }),
    })
    if (!r.ok) throw new Error(`backend respondió ${r.status}`)
    return (await r.json()) as { creadas: number }
  } catch {
    return null
  }
}
